### PR TITLE
Separate model environ upgrades from controller upgrades

### PR DESCRIPTION
**Pull request by axw**
## Description of change

In 2.2 we introduced upgrade steps that would attempt to open an
Environ for every model, and run upgrade operations on the Environ
as necessary (e.g. to move VMs to folders in vSphere). If the
Environ cannot be opened, then the entire controller upgrade fails.

We now introduce a "modelupgrader" worker, which is responsible
for running controller-side upgrade operations for a model's
Environ. Other model workers must wait for this worker to complete
before they can start. If an Environ cannot be opened, or the
Environ's upgrade operations fail, only that model will be
affected.

(Requires https://github.com/juju/juju/pull/7599)

## QA steps

1. juju bootstrap azure azure1 (with 2.1.x)
2. juju add-model foo
(confirm there's no "common" deployment via the portal)
3. juju bootstrap azure azure2 (with this branch)
4. juju switch azure1
5. juju migrate foo azure2
(confirm there's now a "common" deployment in the portal)

The above shows that model upgrade steps are run when a model is migrated to a newer controller.

Repeat, using "juju upgrade-juju" on the original controller to show that the same works with an upgrade.

Repeat with "juju upgrade-juju", but this time inject a failure (before running the upgrade-juju command):
```
mkdir /var/lib/juju/wrench
cat fail-all > /var/lib/juju/wrench/modelupgrader
```

And confirm that the controller is upgraded, and only the model workers are broken until the wrench file is removed.

## Documentation changes

None.

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1700451

### CODE REVIEW

File: cmd/jujud/agent/model/manifolds.go [hunk 1]
```diff
@@ -159,23 +191,23 @@ func Manifolds(config ManifoldsConfig) dependency.Manifolds {
 		// fortress).
 		//
 		// Note that the fortress and flag will only exist while
-		// the model is not dead; this frees their dependencies
-		// from model-lifetime concerns.
-		migrationFortressName: ifNotDead(fortress.Manifold()),
-		migrationInactiveFlagName: ifNotDead(migrationflag.Manifold(migrationflag.ManifoldConfig{
+		// the model is not dead, and not upgrading; this frees
+		// their dependencies from model-lifetime/upgrade concerns.
+		migrationFortressName: ifNotUpgrading(ifNotDead(fortress.Manifold())),
```
Comment (babbageclunk) [inline cmd/jujud/agent/model/manifolds.go]: Is it worth combining `ifNotUpgrading(ifNotDead(` into one decorator? I find this a bit hard to understand.

File: upgrades/environ.go [hunk 2]
```diff
@@ -1,76 +0,0 @@
-// Copyright 2017 Canonical Ltd.
```
Comment (babbageclunk) [inline upgrades/environ.go]: Nice!

File: apiserver/modelupgrader/backend.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+func (s stateBackend) GetModel(tag names.ModelTag) (Model, error) {
+	m, err := s.State.GetModel(tag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return stateModel{m}, nil
+}
+
+type stateModel struct {
```
Comment (axw) [inline apiserver/modelupgrader/backend.go]: indeed, dropped

File: apiserver/modelupgrader/modelupgrader.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	model, err := f.backend.GetModel(tag)
+	if err != nil {
+		return -1, errors.Trace(err)
+	}
+	return model.EnvironVersion(), nil
+}
+
+// ModelTargetEnvironVersion returns the target version of the environ
+// corresponding to each specified model.
```
Comment (axw) [inline apiserver/modelupgrader/modelupgrader.go]: done

File: worker/modelupgrader/manifold_test.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/worker/dependency"
+	dt "github.com/juju/juju/worker/dependency/testing"
+	"github.com/juju/juju/worker/gate"
+	"github.com/juju/juju/worker/modelupgrader"
+)
+
+type ManifoldSuite struct {
```
Comment (axw) [inline worker/modelupgrader/manifold_test.go]: good call, done

File: worker/modelupgrader/worker.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+// newUpgradeWorker returns a worker that runs the upgrade steps, updates the
+// model's environ version, and unlocks the gate.
+func newUpgradeWorker(config Config, targetVersion int) (worker.Worker, error) {
+	currentVersion, err := config.Facade.ModelEnvironVersion(config.ModelTag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return jujuworker.NewSimpleWorker(func(<-chan struct{}) error {
```
Comment (axw) [inline worker/modelupgrader/worker.go]: Only if we make the upgrade steps interruptible. They're not supposed to be long-running things anyway, so I don't think it's necessary.

File: worker/modelupgrader/worker.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+			logger.Tracef(
+				"ignoring upgrade operation for version %v",
+				op.TargetVersion,
+			)
+			continue
+		}
+		if op.TargetVersion > targetVersion {
+			// The operation is for a version newer than
+			// the provider's current version. This will
+			// only happen for an improperlt written provider.
```
Comment (axw) [inline worker/modelupgrader/worker.go]: fixed

File: cmd/jujud/agent/model/manifolds.go (hunk 1 above)
Comment (axw) [inline cmd/jujud/agent/model/manifolds.go]: I don't see a nice, quick win here. I'm going to leave this for now, if you have ideas on how to make it more understandable I'd be happy to hear them.

File: worker/modelupgrader/worker.go (hunk 6 above)
Comment (axw) [inline worker/modelupgrader/worker.go]: I've left a comment here anyway.

**Comment by axw** [comment]
$$merge$$

**Pull request merged**